tutorials/multi/1_graph_classification.py

```diff
@@ -0,0 +1,241 @@
+"""
+Single Machine Multi-GPU Minibatch Graph Classification
+=======================================================
+
+In this tutorial, you will learn how to use multiple GPUs in training a
+graph neural network (GNN) for graph classification. This tutorial assumes
+expertise in GNNs for graph classification and we recommend you to check
+:doc:`Training a GNN for Graph Classification <../blitz/5_graph_classification>` otherwise.
+
+(Time estimate: 8 minutes)
+
+To use a single GPU in training a GNN, we need to put the model, graph(s), and other
+tensors (e.g. labels) on the same GPU:
+"""
+
+"""
+import torch
+
+# Use the first GPU
+device = torch.device("cuda:0")
+model = model.to(device)
+graph = graph.to(device)
+labels = labels.to(device)
+"""
+
+###############################################################################
+# The node and edge features in the graphs, if any, will also be on the GPU.
+# After that, the forward computation, backward computation and parameter
+# update will take place on the GPU. For graph classification, this repeats
+# for each minibatch gradient descent.
+#
+# Using multiple GPUs allows performing more computation per unit of time. It
+# is like having a team work together, where each GPU is a team member. We need
+# to distribute the computation workload across GPUs and let them synchronize
+# the efforts regularly. PyTorch provides convenient APIs for this task with
+# multiple processes, one per GPU, and we can use them in conjunction with DGL.
+#
+# Intuitively, we can distribute the workload along the dimension of data. This
+# allows multiple GPUs to perform the forward and backward computation of
+# multiple gradient descents in parallel. To distribute a dataset across
+# multiple GPUs, we need to partition it into multiple mutually exclusive
+# subsets of a similar size, one per GPU. We need to repeat the random
+# partition every epoch to guarantee randomness. We can use
+# :func:`~dgl.dataloading.pytorch.GraphDataLoader`, which wraps some PyTorch 
+# APIs and does the job for graph classification in data loading.
+#
+# Once all GPUs have finished the backward computation for its minibatch,
+# we need to synchronize the model parameter update across them. Specifically,
+# this involves collecting gradients from all GPUs, averaging them and updating
+# the model parameters on each GPU. We can wrap a PyTorch model with
+# :func:`~torch.nn.parallel.DistributedDataParallel` so that the model
+# parameter update will invoke gradient synchronization first under the hood.
```

Comment:
What does `first` mean here?

Reply:
before actually updating the model parameters